src/GraphQL/Types/QueryArguments.cs

```diff
@@ -49,7 +49,8 @@ public void Add(QueryArgument argument)
             if (argument == null)
                 throw new ArgumentNullException(nameof(argument));
 
-            NameValidator.ValidateName(argument.Name, "argument");
+            if (string.IsNullOrEmpty(argument.Name))
```

Comment:
?

Reply:
The only check left here (which I can move to the setter) is a null/empty check. Ditto everywhere else. The name can’t be tested as it hasn’t yet had the NameConverter applied

Reply:
> The name can’t be tested as it hasn’t yet had the NameConverter applied

Why? I see no reason not to test name. Really.

Reply:
See summary of issue #1840 . The goal is to allow invalid names for a field or argument and specify a name converter that patches them up. Validation occurs after, and not before, the name converter has been applied. So if the schema has names like “my-field” they could be globally translated to “myField”

Reply:
Ah, I completely forgot about it. OK, but let's do this in another way. I propose to move check in setters but allow to `overwrite` it with user supplied one:
```c#
public static class NameValidator
    {
        private const string RESERVED_PREFIX = "__";
        private const string NAME_RX = @"^[_A-Za-z][_0-9A-Za-z]*$";

        public static Action<string, string> NameValidation = ValidateName;  // <---------------------------- ADDED, use it everywhere across repo instead the call to static ValidateName

        public static void ValidateName(string name, string type = "field")
        {
            if (string.IsNullOrWhiteSpace(name))
            {
                throw new ArgumentOutOfRangeException(nameof(name),
                    $"A {type} name can not be null or empty.");
            }

            if (name.Length > 1 && name.StartsWith(RESERVED_PREFIX, StringComparison.InvariantCulture))
            {
                throw new ArgumentOutOfRangeException(nameof(name),
                    $"A {type} name: {name} must not begin with \"__\", which is reserved by GraphQL introspection.");
            }
            if (!Regex.IsMatch(name, NAME_RX))
            {
                throw new ArgumentOutOfRangeException(nameof(name),
                    $"A {type} name must match /^[_a-zA-Z][_a-zA-Z0-9]*$/ but {name} does not.");
            }
        }
    }
```

Reply:
It will be backward compatible change :)

Reply:
The proposed change is already backwards compatible and does not introduce new api. Any code that wouldn’t run before still won’t run at all.  Any code that ran before will still run correctly. 

I think the library should enforce a valid schema. Rewriting the name validation check doesn’t seem to cover that. 

Perhaps if this new property was used by the setters and the built in method was used by the schema initialization (and type name setters, etc) — but that seems confusing.

Reply:
Or a simple static Boolean to skip pre-initialization name checks.

Reply:
> The proposed change is already backwards compatible and does not introduce new api. Any code that wouldn’t run before still won’t run at all. Any code that ran before will still run correctly

With the exception of a **bug** fix: NameConverters could have been used to generate invalid names inside a schema. Now they cannot.

Reply:
> I think the library should enforce a valid schema. 

Yes. But now (and even after your edits) this will not be the case. In addition, we have already come to the understanding that we both want to make it possible to use non-standard (actually invalid) names if someone **really** needs it. And since it is so that it should still be possible to bypass the check somewhere and somehow. That is why I propose not to copy this check throughout the code, which is fraught with errors when one more place for setting the property is added, but to place it in one place. Actually not event in one place since there are many types with name (`INamedType` interface). Now I see check for `type`, `argument`, `enum`. That is, it is still possible to set an invalid name for the directive, for example.

I am trying to convey my idea that it is pointless to try to solve only the local problem indicated in the issue. We need to look at it more broadly and do a centralized check for all named types **inside** these types. No need to scatter checks around codebase. It's error prone, it's hard to maintain.

Reply:
> With the exception of a bug fix: NameConverters could have been used to generate invalid names inside a schema. Now they cannot.

I see no point in it. Why to disallow set invalid names by NameConverters if we still allow to set invalid names from anywhere else ?

Reply:
> We need to look at it more broadly and do a centralized check for all named types inside these types.

I think you need to re-read my code changes again, since this is almost exactly what this PR does.  Now, and unlike previously, the field and argument names are checked in **only one spot** with this PR: `GraphTypesLookup.HandleField`.  It does not matter if you set it in a field builder, or if it's set before or after you've added it to the schema.  It gets checked long after all that's done, when you run your first query against the schema.

I did leave a few null/empty checks in there.  It calls `NameValidatior.ValidateName` only to produce a consistent error message, not because it's actually checking anything.  I can consolidate those checks into the setters.  But they are simply 'bonus checks', and could be completely removed.

Reply:
I have read it several times and I understand what you are talking about. You are talking about single validation when building the schema. And I'm talking about checking regardless of whether schema was built or not. I also say that moving check into setters will make the code **cleaner**. I really don't like the fact that checks are now scattered throughout the code.

> only to produce a consistent error message

I guessed it, but it was not obvious.

Reply:
> I also say that moving check into setters will make the code cleaner. 

I agree, but we must consider that `Name` may never be set

Reply:
> And I'm talking about checking regardless of whether schema was built or not. 

Which is the point of this PR: so that it does not check before the schema is built.

Reply:
> I agree, but we must consider that Name may never be set

This can be checked when initializing the schema. See `SchemaBuilder.Validate` and comment inside. I've wanted to do something like server-side schema validation for a long time (similar to request validation). It is another issue and placing basic checks inside name setters does not conflict with this.

Reply:
> In addition, we have already come to the understanding that we both want to make it possible to use non-standard (actually invalid) names if someone **really** needs it.

If we did, that is beyond the scope of this PR.  This PR should **not** allow invalid names just because someone wants to use a custom `NameConverter`.